cornell-dti/idol-backend
# Refactor data types and api response types

### Summary 





This pull request is the first step to refactor IDOL endpoints for member and teams APIs by defining all data types in `DataTypes.d.ts` and all API response types in `APITypes.d.ts`. The endpoints return the status code and either the error message or the data on success. The error messages and response codes are also updated as per [HTTP response status codes](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status). 

**Error codes for members:**
200 - successful
401 - session user doesn't exist in db
404 - member doesn't exist in db
403 - user does not have permission to edit
400 - undefined email
500 - internal errors due to the db (various reasons)

**Error codes for teams:**
200 - successful
404 - members don't exist in db _or_ no team exists with the entered uuid
403 - user does not have permission to edit
400 - malformed teams or members on POST
500 - internal errors due to db (various reasons)


### Test Plan 


Need to test locally (this includes tests from #9):
- check if the user can update their general info correctly
- check if user can only update their information and not another user's (unless their role is admin or lead)
- check if only admins/leads can update the role
- check if only admins/leads can update the first name/last name
- check if only the user themselves or the admins/leads get the member info
- check if all old member/team endpoints work as intended with the new changes to response types


### Notes 


- For the member type, should the fields `role`, `subteam` and `other_subteams` be updated to types `role`, `Team` and `Team[]` respectively, instead of `string`?
- Not sure if this is needed, but we aren't checking if the request body matches the` Member` type in member endpoints.

### Breaking Changes  


This shouldn't affect any frontend implementation, since it just involves refactoring backend code.
